### ISSUE TITLE
Fela inline styles with nonce

### ISSUE THREAD

**Issue by KyleBastien** [robinweser/fela#742]
**Type**: Feature

| Packages | Version |
| -------- | ------- |
| fela     | 10.6.x   |

-------------------
#### Description
It looks like Fela adds inline styles to the document, that look like this:

https://github.com/robinweser/fela/blob/master/packages/fela-dom/src/server/createStyleTagMarkup.js

But it doesn't look like there is a way to set a nonce for these style tags? Is anyone implementing a strict content-security-policy and using Fela? Any workarounds I'm missing?

If there isn't a workaround that I'm missing, Fela should provide a way to set a nonce via some window object, so that users of fela can implement strict content security policies.

#### Proposal

Create a FelaConfig object, that can be read from the window, and if there is a nonce setting on that object, then apply that nonce to all inline styles being generated by fela.

**Code Example**:
```javascript
// User's code....
window.FelaConfig = {
  cspSettings: {
    nonce: '<user-provided-nonce-value>'
  }
};

// Change to Fela...
export default function createStyleTagMarkup(
  css: string,
  type: StyleType,
  media: string = '',
  rehydrationIndex: number = -1,
  support: boolean = false
): string {
  const mediaAttribute = media.length > 0 ? ` media="${media}"` : ''
  const supportAttribute = support ? ' data-fela-support="true"' : ''
  const nonceAttribute = window.FelaConfig && window.FelaConfig.cspSettings && window.FelaConfig.cspSettings.nonce ? $`nonce-${window.FelaConfig.cspSettings.nonce}` : '';

  return `<style type="text/css" data-fela-rehydration="${rehydrationIndex}" data-fela-type="${type}"${supportAttribute}${mediaAttribute}${nonceAttribute}>${css}</style>`
}
```

**Comment by robinweser** [comment]
@HugoGiraudel isn‘t this the same issue you once had?
I‘m pbly adding a style attributes config to fela. Don‘t want any global window stuff in here.

**Comment by KyleMcNutt** [comment]
@robinweser I think a global window might actually be preferred here, even though I agree it's sort of yucky... if a user is using a library that in turn uses fela, unless that style attributes setting is exposed by that library the user has no way to set a nonce, even if fela does support it. Instead, if the user is able to directly effect fela's settings (via something like a global window object...) then they can opt into a stricter CSP, without any change in a library.

Maybe there is a better way to do it than a global window object, but it should be able to be set by someone who is using a Component library that uses fela, even if that user doesn't directly use fela.

**Comment by HugoGiraudel** [comment]
We postponed our strict CSP integration for other reasons, but we did have the problem that client-side injected `style` elements could not be passed a nonce yes. A configuration should easily fix that. :)

**Comment by robinweser** [comment]
Alright I see!
So I'm still more into the idea of having a config here. If a component library uses Fela, I feel like they'd have to expose the settings and/or allow window.whatever. Would you be ok with that @KyleMcNutt?

**Comment by KyleMcNutt** [comment]
@robinweser I would still prefer Fela to take a config object from window directly if possible, or to take both options and prefer the settings object over the window object. It would really help with not having to chase down each Component library and asking them to also expose this setting.

But if you're not open to this being exposed on a window object, I'd be okay with it being exposed via a settings object, at least that would be progress in the right direction.

**Comment by robinweser** [comment]
@KyleMcNutt Would you be willing to implement that behaviour? Pbly support both window and renderer config would be the way to go then :)

**Comment by KyleMcNutt** [comment]
@robinweser I probably won't have the chance to work on this immediately, but I'll try and take this up.

> support both window and renderer config would be the way to go then

Can you point me at other examples of a renderer config in fela? So I get an idea of how to plumb that through to here?

**Comment by robinweser** [comment]
So, those are all options right now: http://fela.js.org/docs/advanced/RendererConfiguration.html
Which are added to the renderer here: https://github.com/robinweser/fela/blob/master/packages/fela/src/createRenderer.js#L44

In fela-dom, you'll have access to the renderer, so you can read from that config (or read from the window.___ object) and add additional attributes.

It should be passed down to this:
https://github.com/robinweser/fela/blob/master/packages/fela-dom/src/server/createStyleTagMarkup.js for fela-dom's `renderToMarkup`.

Maybe added here: https://github.com/robinweser/fela/blob/master/packages/fela-dom/src/server/renderToSheetList.js although this is at users hand anyways, but feels nice to have consistency.

And finally applied here https://github.com/robinweser/fela/blob/master/packages/fela-dom/src/dom/connection/createNode.js which is the runtime code that generates style tags.

Also I'd propose the following API:
```
const attributes = {
  "nonce-XXX": ...
}

// and then either
const renderer = createRenderer({ styleNodeAttributes: attributes })

// or
window.FelaConfig.styleNodeAttributes = attributes
```

This gives us more flexibility and allows for possible future cases where we'd like to add other props as well.

**Comment by robinweser** [comment]
Implemented with #768 and released today (3.3.2020)
